### PR TITLE
Support for basic list comprehensions

### PR DESCRIPTION
**Pull request by Krovatkin**
Supports the following syntax:
```
        @torch.jit.script
        def comp(l):
            # type: (List[float]) -> List[float]

            n = [x * 3 for x in l]
            return n
```

### CODE REVIEW

File: torch/csrc/jit/script/compiler.cpp [hunk 1]
```diff
@@ -955,6 +956,49 @@ struct to_ir {
...
+    auto apply_append =
+        Apply::create(lc.range(), dot_op, append_args_list, append_attrs);
+    auto expr_stmt = ExprStmt::create(lc.range(), apply_append);
+    auto stmt_list = List<Stmt>::create(lc.range(), {expr_stmt});
+    auto iters_list = List<Expr>::create(lc.range(), {lc.iter()});
+    auto targets_list = List<Expr>::create(lc.range(), {lc.target()});
+    auto for_loop =
+        For::create(lc.range(), targets_list, iters_list, stmt_list);
+    emitFor(for_loop);
+    return n->output();
```
Comment (suo) [inline torch/csrc/jit/script/compiler.cpp]: Return the output of the prim::Loop node here instead of the original ListConstruct

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -955,6 +956,49 @@ struct to_ir {
     return emitIfExpr(expr.range(), cond_value, true_expr, false_expr);
   }
 
+  Value* emitListComprehension(const ListComp& lc) {
```
Comment (suo) [inline torch/csrc/jit/script/compiler.cpp]: I feel like there could be more consts in this function :)

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -955,6 +956,49 @@ struct to_ir {
...
+    // on x
+
+    // given `[x*2 for x in my_list]` this generates the following AST:
+    // __list_acc = []
+    // for x in my_list:
+    //  __list_acc.append(x*2)
+    auto n = graph->insertNode(
+        graph->createList(elem_types.at(0), at::ArrayRef<Value*>{}));
+    environment_stack->setVar(lc.range(), tmp_name, n->output());
+    auto tmp_list_ident = Ident::create(lc.range(), tmp_name);
```
Comment (suo) [inline torch/csrc/jit/script/compiler.cpp]: oof, I agree that an AST rewrite is probably the best way to do list comprehensions, but this code is hard to read. I guess we don't have a better way to write this now :/

File: torch/csrc/jit/script/tree_views.h [hunk 4]
```diff
@@ -646,6 +671,7 @@ struct BinOp : public Expr {
       const Expr& rhs) {
     return BinOp(Compound::create(kind, range, {lhs, rhs}));
   }
+
```
Comment (suo) [inline torch/csrc/jit/script/tree_views.h]: looks like you added some extra lines?

File: test/test_jit.py [hunk 5]
```diff
@@ -3947,6 +3947,53 @@ def test_append():
             return a == [0, 1, 2, 3]
         self.checkScript(test_append, ())
 
+    def test_comprehensions_basic(self):
+
+        @torch.jit.script
```
Comment (suo) [inline test/test_jit.py]: use checkScript here to verify that the both frontends (AST and string) return equivalent results to python

File: torch/csrc/jit/script/tree_views.h [hunk 6]
```diff
@@ -465,6 +466,30 @@ struct For : public Stmt {
...
+  Expr elt() const {
+    return Expr(subtree(0));
+  }
+  Expr target() const {
+    return Expr(subtree(1));
+  }
+  Expr iter() const {
+    return Expr(subtree(2));
+  }
+  // TODO: no ifs for now
```
Comment (suo) [inline torch/csrc/jit/script/tree_views.h]: What does the error message look like if the user uses one of the unsupported features?

File: test/test_jit.py [hunk 7]
```diff
@@ -3947,6 +3947,53 @@ def test_append():
...
+            # type: (List[int], List[int]) -> List[int]
+
+            n = [x * 3 for x in l1]
+            n2 = [x + 2 for x in l2]
+            return n + n2
+
+        self.assertEqual(comp([1, 2, 3], [4, 5]), [3, 6, 9, 6, 7])
+
+    def test_comprehensions_wrong_expr_type(self):
+
```
Comment (suo) [inline test/test_jit.py]: formatting nit: generally we don't leave blank lines after `def`, `with` or type annotations

**Review by suo — COMMENTED** [review]
Nice! You'll have to add support for list comprehensions from our string frontend (check parser.h). the `checkScript` test function will help with that.

File: torch/csrc/jit/script/compiler.cpp (hunk 1 above)
Comment (suo) [reply torch/csrc/jit/script/compiler.cpp]: This comment is still valid I believe

File: torch/csrc/jit/script/parser.cpp [hunk 8]
```diff
@@ -385,7 +403,8 @@ struct ParserImpl {
     auto rhs = parseExpOrExpTuple();
     L.expect(TK_NEWLINE);
     if (op->kind() == '=') {
-      return Assign::create(lhs.range(), lhs, Expr(rhs));
+      auto asep = Assign::create(lhs.range(), lhs, Expr(rhs));
```
Comment (suo) [inline torch/csrc/jit/script/parser.cpp]: nit: revert this change?

**Comment by Krovatkin** [comment]
@suo I simplified the check in parser. Please take another look!

**Comment by suo** [comment]
@Krovatkin looks like there are some unaddressed comments?

File: torch/csrc/jit/script/compiler.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/script/compiler.cpp]: `emitFor` doesn't return a `Value` and `%8` isn't automatically added to loop outputs. Neither one is insurmountable but it will add even **more** boilerplate.

```
graph(%l : int[]):
  %1 : bool = prim::Constant[value=1]()
  %2 : int = prim::Constant[value=3]()
  %___list_acc0 : int[] = prim::ListConstruct()
  %4 : int = aten::len(%l)
   = prim::Loop(%4, %1)
    block0(%5 : int):
      %x : int = aten::select(%l, %5)
      %7 : int = aten::mul(%x, %2)
      %8 : int[] = aten::append(%___list_acc0, %7)
      -> (%1)
  return (%___list_acc0)
```

**Review by suo — APPROVED** [review]
lgtm
